cpp/src/text/minhash.cu

```diff
@@ -0,0 +1,188 @@
+  cudf::size_type width;
+  cudf::hash_value_type* d_hashes;
+
+  __device__ void operator()(cudf::size_type idx)
+  {
+    auto const str_idx  = idx / cudf::detail::warp_size;
+    auto const lane_idx = idx % cudf::detail::warp_size;
+
+    if (d_strings.is_null(str_idx)) { return; }
+
+    auto const d_str    = d_strings.element<cudf::string_view>(str_idx);
+    auto const d_output = d_hashes + (str_idx * seeds.size());
+
+    // initialize hashes output for this string
+    if (lane_idx == 0) {
+      auto const init = d_str.empty() ? 0 : std::numeric_limits<cudf::hash_value_type>::max();
+      thrust::fill(thrust::seq, d_output, d_output + seeds.size(), init);
+    }
+    __syncwarp();
+
+    auto const begin = d_str.begin() + lane_idx;
+    auto const end   = [d_str, width = width] {
+      auto const length = d_str.length();
+      if (length > width) { return (d_str.end() - (width - 1)); }
+      return d_str.begin() + static_cast<cudf::size_type>(length > 0);
+    }();
+
+    // each lane hashes substrings of the given width
+    for (auto itr = begin; itr < end; itr += cudf::detail::warp_size) {
+      auto const offset = itr.byte_offset();
+      auto const hash_str =
+        cudf::string_view(d_str.data() + offset, (itr + width).byte_offset() - offset);
+
+      // hashing each seed on the same section of string is 10x faster than
+      // computing the substrings for each seed
+      for (std::size_t seed_idx = 0; seed_idx < seeds.size(); ++seed_idx) {
+        auto const hasher = cudf::detail::MurmurHash3_32<cudf::string_view>{seeds[seed_idx]};
+        auto const hvalue = hasher(hash_str);
+        atomicMin(d_output + seed_idx, hvalue);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<cudf::column> minhash(cudf::strings_column_view const& input,
+                                      cudf::device_span<cudf::hash_value_type const> seeds,
+                                      cudf::size_type width,
+                                      cudf::hash_id hash_function,
+                                      rmm::cuda_stream_view stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!seeds.empty(), "Parameter seeds cannot be empty", std::invalid_argument);
+  CUDF_EXPECTS(width >= 2,
+               "Parameter width should be an integer value of 2 or greater",
+               std::invalid_argument);
+  CUDF_EXPECTS(hash_function == cudf::hash_id::HASH_MURMUR3,
+               "Only murmur3 hash algorithm supported",
+               std::invalid_argument);
+
+  auto output_type = cudf::data_type{cudf::type_to_id<cudf::hash_value_type>()};
+  if (input.is_empty()) { return cudf::make_empty_column(output_type); }
+
+  auto const d_strings = cudf::column_device_view::create(input.parent(), stream);
+
+  auto hashes   = cudf::make_numeric_column(output_type,
+                                          input.size() * static_cast<cudf::size_type>(seeds.size()),
```

Comment:
Could this multiplication overflow, and do we care (either here, or in the python wrapper)?

Reply:
Yes. You are correct. This could technically overflow. 
This would mean we are given a large number of very small strings which is not very likely here but certainly possible. 
I will add a check for this.

Reply:
Yes. Added a check for this and a gtest.